File: backend/src/main/java/com/mycompany/online_shop_backend/controllers/OrderController.java
package com.mycompany.online_shop_backend.controllers;

import com.mycompany.online_shop_backend.dto.request.OrderRequest;
import com.mycompany.online_shop_backend.dto.response.CreatedOrderResponse;
import com.mycompany.online_shop_backend.dto.response.UserOrderDto;
import com.mycompany.online_shop_backend.dto.services.OrderDto;
import com.mycompany.online_shop_backend.services.OrderService;
import com.mycompany.online_shop_backend.services.SecurityService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.web.bind.annotation.*;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

@RestController
@RequiredArgsConstructor
public class OrderController {

    private final OrderService orderService;
    private final SecurityService securityService;

    @PostMapping(
            path = "/api/v1/orders",
            consumes = MediaType.APPLICATION_JSON_VALUE,
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    @ResponseStatus(HttpStatus.CREATED)
    public CreatedOrderResponse createOrder(@RequestBody OrderRequest request) {
        OrderDto orderDto = orderService.save(request);
        return CreatedOrderResponse.toDto(orderDto);
    }

    @GetMapping(
            path = "/api/v1/users/{id}/orders",
            produces = MediaType.APPLICATION_JSON_VALUE
    )
    public List<UserOrderDto> getUserOrders(HttpServletRequest request) {
        String email = securityService.getUsernameFromRequest(request);
        List<OrderDto> orderDtos = orderService.getOrdersByEmail(email);
        return orderDtos.stream().map(UserOrderDto::toDto).toList();
    }
}

File: 04-services/order-service/src/main/kotlin/com/example/orderservice/api/rest/OrderController.kt
package com.example.orderservice.api.rest

import com.example.orderservice.api.rest.model.*
import com.example.orderservice.service.OrderService
import io.swagger.v3.oas.annotations.Operation
import org.springframework.data.web.PagedModel
import org.springframework.http.HttpStatus
import org.springframework.http.MediaType
import org.springframework.web.bind.annotation.*
import java.util.*

@RestController
@RequestMapping("/api/v1", produces = [MediaType.APPLICATION_JSON_VALUE])
class OrderController(private val orderService: OrderService) {

    @Operation(summary = "Search orders")
    @PostMapping("/orders/search", consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun orders(
        orderRequestParams: OrderRequestParams,
        @RequestBody request: OrderSearchRequest
    ): PagedModel<Order> = orderService.getOrders(orderRequestParams, request)

    @Operation(summary = "Get order by id")
    @GetMapping("/orders/{orderId}")
    fun orderById(@PathVariable("orderId") orderId: UUID): Order = orderService.getOrderById(orderId)

    @Operation(summary = "Create order")
    @PostMapping("/orders", consumes = [MediaType.APPLICATION_JSON_VALUE])
    @ResponseStatus(HttpStatus.CREATED)
    fun createOrder(
        @RequestHeader(IDEMPOTENCY_KEY) idempotencyKey: UUID,
        @RequestBody request: CreateOrderRequest
    ): CreateOrderResponse = orderService.createOrder(idempotencyKey, request)

    @Operation(summary = "Update order status")
    @PatchMapping("/orders/{orderId}/status", consumes = [MediaType.APPLICATION_JSON_VALUE])
    fun updateOrderStatus(
        @PathVariable("orderId") orderId: UUID,
        @RequestBody request: UpdateOrderStatusRequest
    ): Order = orderService.updateOrderStatus(orderId, request.status)

}